k2/python/k2/rnnt_loss.py

```diff
@@ -1159,6 +1235,24 @@ def rnnt_loss_smoothed(
         boundary=boundary,
         modified=modified,
     )
+
+    if delay_penalty > 0.0:
+        B, S, T0 = px.shape
+        T = T0 if modified else T0 - 1
+        if boundary is None:
+            offset = torch.tensor(
+                [(T - 1) / 2 * delay_penalty] * B,
+                dtype=px.dtype,
+                device=px.device,
+            )
+        else:
+            offset = (boundary[:, 3] - 1) / 2 * delay_penalty
+        penalty = offset.reshape(B, 1, 1) - delay_penalty * torch.arange(
+            T0, device=px.device
+        ).reshape(1, 1, T0)
+        penalty = penalty.to(px.dtype)
+        px += penalty
+
     scores_and_grads = mutual_information_recursion(
         px=px, py=py, boundary=boundary, return_grad=return_grad
```

Comment:
I think it might be useful to keep track of the average delay during training, which we could get
by computing the dot-product of the (arange - offset) with the returned grads.  
But I'm not sure that it would be super convenient to add an extra return value to these
functions because it would require the calling code to be changed at the exact same time.
It would definitely be possible to also compute it outside this function, in icefall, by duplicating the
code that computes the (arange - offset) and dotting that with the returned grad.  I'd say that
would be an acceptable solution.
That would make it obvious if we were penalizing the delay too much or too little; it would also
provide an easy way to measure what the delay "should" be as a baseline, if we ran non-chunked
systems.  (I.e. I'm saying we could perhaps keep track of this avg-delay even in cases when we were not
doing the delay-penalty).

Reply:
Yes, I think so. I kept track of the average delay in Icefall, in the model.py and return it along loss value. I have not created a PR yet.